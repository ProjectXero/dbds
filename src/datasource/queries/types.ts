import { SqlSqlTokenType } from "slonik"

export type PrimitiveValueType = string | number | boolean | null
export type ValueType =
  | PrimitiveValueType
  | Array<PrimitiveValueType>
  | ReadonlyArray<PrimitiveValueType>

export type ConditionValue<T extends string | number | boolean> = T | null | Array<T | null>

export type GenericConditions = Record<string, ConditionValue<string> | ConditionValue<number> | ConditionValue<boolean> | SqlSqlTokenType>

export type Conditions<TRowType> = {
  [K in keyof TRowType]?: TRowType[K] extends string
  ? ConditionValue<string>
  : TRowType[K] extends number
  ? ConditionValue<number>
  : TRowType[K] extends boolean
  ? ConditionValue<boolean>
  : SqlSqlTokenType
} & GenericConditions
