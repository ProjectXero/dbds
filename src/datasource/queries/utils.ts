import {
  IdentifierSqlToken,
  FragmentSqlToken,
  isSqlToken as isSlonikToken,
  SqlToken,
} from 'slonik'
import { FragmentToken } from 'slonik/dist/src/tokens'
import { OrderTuple } from './types'

export const isSqlToken = (subject: unknown): subject is SqlToken => {
  return isSlonikToken(subject)
}

export const isFragmentSqlToken = (v: unknown): v is FragmentSqlToken => {
  return (
    typeof v === 'object' &&
    Object.prototype.hasOwnProperty.call(v, 'sql') &&
    Object.prototype.hasOwnProperty.call(v, 'type') &&
    Object.prototype.hasOwnProperty.call(v, 'values') &&
    (v as FragmentSqlToken).type === FragmentToken
  )
}

export const isIdentifierSqlToken = (v: unknown): v is IdentifierSqlToken => {
  return (
    typeof v === 'object' &&
    Object.prototype.hasOwnProperty.call(v, 'names') &&
    Object.prototype.hasOwnProperty.call(v, 'type')
  )
}

export const isOrderTuple = (v: unknown): v is OrderTuple => {
  return (
    Array.isArray(v) &&
    v.length >= 1 &&
    v.length <= 2 &&
    (v[1] === 'ASC' || v[1] === 'DESC') &&
    (typeof v[0] === 'string' ||
      isIdentifierSqlToken(v[0]) ||
      isFragmentSqlToken(v[0]))
  )
}
