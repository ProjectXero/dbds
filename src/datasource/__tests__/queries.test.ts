import { sql } from 'slonik'
import { QueryBuilder } from '../queries'

interface DummyRowType {
  id: number
  name: string
  optional?: string
  nullable: string | null
  optionallyNullable?: string | null
  stringOrNumber: string | number | null // i think this is a nonsense type but i needed to test the type system...
}

const DummyRowColumnTypes = Object.freeze({
  id: 'these',
  name: 'types',
  optional: 'do',
  nullable: 'not',
  optionallyNullable: 'matter',
  stringOrNumber: 'here',
})

describe(QueryBuilder, () => {
  const builder = new QueryBuilder<DummyRowType>(
    'any_table',
    DummyRowColumnTypes
  )

  describe('clause generators', () => {
    describe('where', () => {
      it('accepts a simple object', () => {
        expect(builder.where({ id: 1 })).toMatchSnapshot()
      })

      it('accepts complex conditions', () => {
        expect(
          builder.where({ id: 1, nullable: null, stringOrNumber: ['a', null] })
        ).toMatchSnapshot()
      })

      it('uses AND for multiple columns in a simple object', () => {
        expect(builder.where({ id: 1, name: 'Bob' })).toMatchSnapshot()
      })

      it('handles multiple values for a column', () => {
        expect(builder.where({ id: [1, 2, 3] })).toMatchSnapshot()
      })

      it('lets you pass in a raw expression', () => {
        expect(builder.where(sql`true`)).toMatchSnapshot()
      })

      it('accepts a list of expressions', () => {
        expect(
          builder.where([sql`true`, sql`false`, sql`more raw expressions etc.`])
        ).toMatchSnapshot()
      })

      it('sanely handles output from the `and` and `or` utilities', () => {
        const conditions = builder.or([
          builder.and({ id: 1, name: 'asdf' }),
          builder.and({ id: 2, name: 'wow' }),
        ])

        expect(builder.where(conditions)).toMatchSnapshot()
      })

      it('produces a valid clause with no conditions', () => {
        expect(builder.where({})).toMatchSnapshot()
        expect(builder.where([])).toMatchSnapshot()
      })
    })

    describe('orderBy', () => {
      it('accepts a list of column names as strings', () => {
        expect(builder.orderBy('a', 'b', 'c')).toMatchSnapshot()
      })

      it('accepts a list of column names as an array of strings', () => {
        expect(builder.orderBy(['a', 'b', 'c'])).toMatchSnapshot()
      })

      it('accepts identifier tokens', () => {
        expect(builder.orderBy(sql.identifier(['column']))).toMatchSnapshot()
      })

      it('accepts arbitrary sql tokens', () => {
        expect(builder.orderBy(sql`anything i want!`)).toMatchSnapshot()
      })

      it('accepts a mix of different types', () => {
        expect(
          builder.orderBy('a', sql.identifier(['b']), sql`c`, [
            'd',
            sql.identifier(['e']),
            sql`f`,
          ])
        ).toMatchSnapshot()
      })
    })

    describe('groupBy', () => {
      it('accepts a list of column names as strings', () => {
        expect(builder.groupBy('a', 'b', 'c')).toMatchSnapshot()
      })

      it('accepts a list of column names as an array of strings', () => {
        expect(builder.groupBy(['a', 'b', 'c'])).toMatchSnapshot()
      })

      it('accepts identifier tokens', () => {
        expect(builder.groupBy(sql.identifier(['column']))).toMatchSnapshot()
      })

      it('accepts arbitrary sql tokens', () => {
        expect(builder.groupBy(sql`anything i want!`)).toMatchSnapshot()
      })

      it('accepts a mix of different types', () => {
        expect(
          builder.groupBy('a', sql.identifier(['b']), sql`c`, [
            'd',
            sql.identifier(['e']),
            sql`f`,
          ])
        ).toMatchSnapshot()
      })
    })

    describe('having', () => {
      it('accepts a simple object', () => {
        expect(builder.having({ id: 1 })).toMatchSnapshot()
      })

      it('uses AND for multiple columns in a simple object', () => {
        expect(builder.having({ id: 1, name: 'Bob' })).toMatchSnapshot()
      })

      it('handles multiple values for a column', () => {
        expect(builder.having({ id: [1, 2, 3] })).toMatchSnapshot()
      })

      it('lets you pass in a raw expression', () => {
        expect(builder.having(sql`true`)).toMatchSnapshot()
      })

      it('accepts a list of expressions', () => {
        expect(
          builder.having([
            sql`true`,
            sql`false`,
            sql`more raw expressions etc.`,
          ])
        ).toMatchSnapshot()
      })

      it('sanely handles output from the `and` and `or` utilities', () => {
        const conditions = builder.or([
          builder.and({ id: 1, name: 'asdf' }),
          builder.and({ id: 2, name: 'wow' }),
        ])

        expect(builder.having(conditions)).toMatchSnapshot()
      })

      it('produces a valid clause with no conditions', () => {
        expect(builder.having({})).toMatchSnapshot()
        expect(builder.having([])).toMatchSnapshot()
      })
    })
  })

  describe('core query builders', () => {
    describe('select', () => {
      it('selects everything by default', () => {
        expect(builder.select()).toMatchSnapshot()
      })
    })

    describe('count', () => {
      it('creates a count query', () => {
        expect(builder.count()).toMatchSnapshot()
      })

      it('can use where clauses', () => {
        expect(builder.count({ where: { id: 1 } })).toMatchSnapshot()
      })

      it('does not support group by clauses for the moment', () => {
        expect(() =>
          builder.count({ groupBy: 'id' })
        ).toThrowErrorMatchingInlineSnapshot(
          `"count does not currently support GROUP BY clauses"`
        )
      })
    })

    describe('insert', () => {
      it('accepts a basic object', () => {
        expect(
          builder.insert({
            id: 1,
            name: 'name',
            nullable: null,
            stringOrNumber: 1,
          })
        ).toMatchSnapshot()
      })

      it('accepts many basic objects', () => {
        expect(
          builder.insert([
            {
              id: 1,
              name: 'name',
              nullable: null,
              stringOrNumber: 1,
            },
            {
              stringOrNumber: 'wat',
              id: 2,
              name: 'name',
              nullable: 'hi',
            },
          ])
        ).toMatchSnapshot()
      })
    })

    describe('update', () => {
      it('accepts a basic object', () => {
        expect(
          builder.update({
            id: 5,
            name: 'any',
            nullable: null,
            optional: 'asdf',
            optionallyNullable: null,
            stringOrNumber: 5,
          })
        ).toMatchSnapshot()
      })

      it('accepts raw sql values', () => {
        expect(builder.update({ name: sql`anything i want` })).toMatchSnapshot()
      })
    })

    describe('delete', () => {
      it("doesn't let you delete everything without explicitly okaying it", () => {
        // @ts-expect-error
        expect(() => builder.delete()).toThrowErrorMatchingInlineSnapshot(
          `"Implicit deletion of everything is not allowed. To delete everything, please pass \`true\` or include options."`
        )
      })

      it('can be forced to delete everything', () => {
        expect(builder.delete(true)).toMatchSnapshot()
      })

      it('builds clauses correctly', () => {
        expect(
          builder.delete({
            where: { id: 1 },
            groupBy: 'id',
            orderBy: 'id',
            having: { id: 1 },
          })
        ).toMatchSnapshot()
      })
    })
  })
})
