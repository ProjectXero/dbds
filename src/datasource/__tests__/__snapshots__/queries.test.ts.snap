// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilder clause generators groupBy accepts a list of column names as an array of strings 1`] = `
Object {
  "sql": "GROUP BY \\"any_table\\".\\"a\\", \\"any_table\\".\\"b\\", \\"any_table\\".\\"c\\"",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators groupBy accepts a list of column names as strings 1`] = `
Object {
  "sql": "GROUP BY \\"any_table\\".\\"a\\", \\"any_table\\".\\"b\\", \\"any_table\\".\\"c\\"",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators groupBy accepts a mix of different types 1`] = `
Object {
  "sql": "GROUP BY \\"any_table\\".\\"a\\", \\"b\\", c, \\"any_table\\".\\"d\\", \\"e\\", f",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators groupBy accepts arbitrary sql tokens 1`] = `
Object {
  "sql": "GROUP BY anything i want!",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators groupBy accepts identifier tokens 1`] = `
Object {
  "sql": "GROUP BY \\"column\\"",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators having accepts a list of expressions 1`] = `
Object {
  "sql": "HAVING (true AND false AND more raw expressions etc.)",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators having accepts a simple object 1`] = `
Object {
  "sql": "HAVING (\\"any_table\\".\\"id\\" = $1)",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    1,
  ],
}
`;

exports[`QueryBuilder clause generators having handles multiple values for a column 1`] = `
Object {
  "sql": "HAVING (\\"any_table\\".\\"id\\" = ANY($1::integer[]))",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    Array [
      1,
      2,
      3,
    ],
  ],
}
`;

exports[`QueryBuilder clause generators having lets you pass in a raw expression 1`] = `
Object {
  "sql": "HAVING true",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators having produces a valid clause with no conditions 1`] = `
Object {
  "sql": "HAVING true",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators having produces a valid clause with no conditions 2`] = `
Object {
  "sql": "HAVING true",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators having sanely handles output from the \`and\` and \`or\` utilities 1`] = `
Object {
  "sql": "HAVING ((\\"any_table\\".\\"id\\" = $1 AND \\"any_table\\".\\"name\\" = $2) OR (\\"any_table\\".\\"id\\" = $3 AND \\"any_table\\".\\"name\\" = $4))",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    1,
    "asdf",
    2,
    "wow",
  ],
}
`;

exports[`QueryBuilder clause generators having uses AND for multiple columns in a simple object 1`] = `
Object {
  "sql": "HAVING (\\"any_table\\".\\"id\\" = $1 AND \\"any_table\\".\\"name\\" = $2)",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    1,
    "Bob",
  ],
}
`;

exports[`QueryBuilder clause generators orderBy accepts a list of column names as an array of strings 1`] = `
Object {
  "sql": "ORDER BY \\"any_table\\".\\"a\\", \\"any_table\\".\\"b\\", \\"any_table\\".\\"c\\"",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators orderBy accepts a list of column names as strings 1`] = `
Object {
  "sql": "ORDER BY \\"any_table\\".\\"a\\", \\"any_table\\".\\"b\\", \\"any_table\\".\\"c\\"",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators orderBy accepts a mix of different types 1`] = `
Object {
  "sql": "ORDER BY \\"any_table\\".\\"a\\", \\"b\\", c, \\"any_table\\".\\"d\\", \\"e\\", f",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators orderBy accepts arbitrary sql tokens 1`] = `
Object {
  "sql": "ORDER BY anything i want!",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators orderBy accepts identifier tokens 1`] = `
Object {
  "sql": "ORDER BY \\"column\\"",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators where accepts a list of expressions 1`] = `
Object {
  "sql": "WHERE (true AND false AND more raw expressions etc.)",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators where accepts a simple object 1`] = `
Object {
  "sql": "WHERE (\\"any_table\\".\\"id\\" = $1)",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    1,
  ],
}
`;

exports[`QueryBuilder clause generators where handles multiple values for a column 1`] = `
Object {
  "sql": "WHERE (\\"any_table\\".\\"id\\" = ANY($1::integer[]))",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    Array [
      1,
      2,
      3,
    ],
  ],
}
`;

exports[`QueryBuilder clause generators where lets you pass in a raw expression 1`] = `
Object {
  "sql": "WHERE true",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators where produces a valid clause with no conditions 1`] = `
Object {
  "sql": "WHERE true",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators where produces a valid clause with no conditions 2`] = `
Object {
  "sql": "WHERE true",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder clause generators where sanely handles output from the \`and\` and \`or\` utilities 1`] = `
Object {
  "sql": "WHERE ((\\"any_table\\".\\"id\\" = $1 AND \\"any_table\\".\\"name\\" = $2) OR (\\"any_table\\".\\"id\\" = $3 AND \\"any_table\\".\\"name\\" = $4))",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    1,
    "asdf",
    2,
    "wow",
  ],
}
`;

exports[`QueryBuilder clause generators where uses AND for multiple columns in a simple object 1`] = `
Object {
  "sql": "WHERE (\\"any_table\\".\\"id\\" = $1 AND \\"any_table\\".\\"name\\" = $2)",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    1,
    "Bob",
  ],
}
`;

exports[`QueryBuilder core query builders delete builds clauses correctly 1`] = `
Object {
  "sql": "
      WITH \\"delete_rows\\" AS (
        
      DELETE FROM \\"any_table\\"
      WHERE (\\"any_table\\".\\"id\\" = $1)
      RETURNING *
    
      ) SELECT *
        FROM \\"delete_rows\\"
        GROUP BY \\"any_table\\".\\"id\\"
        ORDER BY \\"any_table\\".\\"id\\"
    ",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [
    1,
  ],
}
`;

exports[`QueryBuilder core query builders delete can be forced to delete everything 1`] = `
Object {
  "sql": "
      WITH \\"delete_rows\\" AS (
        
      DELETE FROM \\"any_table\\"
      
      RETURNING *
    
      ) SELECT *
        FROM \\"delete_rows\\"
        
        
    ",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;

exports[`QueryBuilder core query builders select selects everything by default 1`] = `
Object {
  "sql": "
      SELECT *
      FROM \\"any_table\\"
      
      
      
    ",
  "type": "SLONIK_TOKEN_SQL",
  "values": Array [],
}
`;
